### ISSUE TITLE
lz4 sources, seems, never been adapted to be amalgamated.

### ISSUE THREAD

**Issue by FrankStain** [lz4/lz4#396]
Hi there. :)

So, to be short. There is some technique exist, which is specialized in robust speedup of project compilation. It is much relevant for huge projects or for projects with extensive automated code generation. This technique is named "amalgamation of source code".
It means the all source code (*.c|*.cpp| ... etc.) files are combined into single file(per project) for compilation.
Scalable [build systems](https://github.com/fffaraz/awesome-cpp#build-systems) use this technique alot.

And here is an issue.
The text of source files `lz4.c` and `lz4hc.c` is much duplicated. More than, `#include "lz4.c"` inside of `lz4hc.c` makes direct code duplication which causes the symbols redefinition.
There are alot of duplicated macro names detween `lz4.c` and `lz4frame.c` files. And some macro (like `ALLOCATOR`) differently declared in `lz4.c` and `lz4hc.c`.
Alot of enumeration types are duplicated between `lz4.c` and `lz4hc.c`.
Alot of constants are duplicated across all source code.

Seems i can't fix this issues lone myself cuz it breaks the further compilation. I definitely don't understand how to solve this issues to proper compilation of amalgamated lz4 code.
Can smbdy help me in this issue?

**Comment by Cyan4973** [comment]
Some of these issues could be solved, like duplicated macro in `lz4frame.c`,
others are more difficult to sort out, if that's even possible.

I guess it's necessary to use the same tool as you do to make sure to experience the same problems, and try to fix them.
Could you describe your transformation process ?

**Comment by FrankStain** [comment]
@Cyan4973 
There are few algorithms available for C++ source code amalgamation. But the most distributed and most dumb way is the source code concatenation. All other algorithms are just more clever versions of this one.

Just like [all-in-one source](https://github.com/google/googletest/blob/master/googletest/src/gtest-all.cc) from [googletests](https://github.com/google/googletest).
Currently i experimenting with build lz4 using my own build tool. It works just like that, just lists the files in directory and concatenate it together in alphabetic ascending order.
The result file is passed to compilation phase.

So, for lz4 project such amalgamation will be same as next source file:
```.cpp
#include "lz4.c"
#include "lz4frame.c"
#include "lz4hc.c"
#include "xxhash.c"
```

BTW. Here is good enough article about amalgamation of source code: :-)
https://blog.forrestthewoods.com/improving-open-source-with-amalgamation-cf293592c5f4

**Comment by Cyan4973** [comment]
OK, so for the time being, you're interested in a limited form of amalgamation, with only `*.c` files coalesced together ?
As a consequence, the `*.h` files must remain available, in order for `#include` to work correctly.

It doesn't really simplify distribution as in Forrest's article, but at least it's simple.

**Comment by FrankStain** [comment]
> so for the time being, you're interested in a limited form of amalgamation, with only *.c files coalesced together ?

Yup. :-)

Currently i found next issues:
* https://github.com/lz4/lz4/blob/dev/lib/lz4frame.c#L53 - `ALLOCATOR`, `FREEMEM` and `MEM_INIT` are conflicting with https://github.com/lz4/lz4/blob/dev/lib/lz4.c#L134 .
* https://github.com/lz4/lz4/blob/dev/lib/lz4frame.c#L146 - `KB`, `MB` and `GB` are conflicting with https://github.com/lz4/lz4/blob/dev/lib/lz4.c#L276 .
* https://github.com/lz4/lz4/blob/dev/lib/lz4hc.c#L64 - causes the symbols duplication of all static functions and constants with https://github.com/lz4/lz4/blob/dev/lib/lz4.c#L385 above the marked line.
* https://github.com/lz4/lz4/blob/dev/lib/lz4.c#L274 - the constant is duplicated with itself from https://github.com/lz4/lz4/blob/dev/lib/lz4hc.c .
* https://github.com/lz4/lz4/blob/dev/lib/lz4hc.c#L238 - enumeration type is duplicated with https://github.com/lz4/lz4/blob/dev/lib/lz4.c#L398 and consists of different items.

That is all i currently found. Mby that is all issues, i had stopped my attempts for now and can't say surely.

**Comment by Cyan4973** [comment]
Since `lz4hc.c` includes a part of `lz4.c` (which acts as an embedded `lz4_internal.h`),
the `cat *.c` technique seems just incompatible with it.

A more evolved amalgamation script would be needed, which at least fixes `#include` statements.

**Comment by FrankStain** [comment]
@Cyan4973 , btw, the thing of including the one source file into another looks a bit strange for me. And probably for someone else, who will read the source files.
What do you think about extruding some code from `lz4.c` into separate internal translation unit (.h+.c or just .h with guarding)?

**Comment by Cyan4973** [comment]
It would add another file, and go against an important objective that `lz4.c` + `lz4.h` are enough to experience LZ4.

**Comment by FrankStain** [comment]
I see. Looking this way the proper amalgamation of entire lz4 library looks more appropriate decision.

For now i had define-guarded all text between 35 and 385 lines of `lz4.c`. It takes down the code duplications.
All duplicated macros was guarded by `#ifndef`.
The `ALLOCATOR` macro at `lz4frame.c` was changed to be same as one from `lz4.c`.
At last, i had renamed the `limitedOutput_directive` at `lz4hc.c` with all its items adding the `hc_` prefix.
The last change causes the change at `lz4opt.h`, where the `hc_limitedOutput_directive` type is used in `LZ4HC_compress_optimal` function.

All this fixes lets me pass the compilation. I look on it like on hard violation, but at last the code compiles. :-)
Personally i can work further now. But in a perspective it would be great to hear that lz4 is already amalgamated or able for amalgamation out of box.

**Comment by Cyan4973** [comment]
> it would be great to hear that lz4 is already amalgamated or able for amalgamation out of box.

I guess it means we would need a script that produces amalgamation from current source code.

**Comment by Leandros** [comment]
Amalgamation can be done on windows with a script like this:

```
@echo off

:: Remove old source
del /F /Q lz4_amalgamated.c

:: Generate amalgamated source
echo: > lz4_amalgamated.c
type file1.c >> lz4_amalgamated.c
echo: > lz4_amalgamated.c
type file2.c >> lz4_amalgamated.c
echo: > lz4_amalgamated.c
:: and so on ...
```

with the equivalent bash script:

```
#!/bin/sh

# Remove old source
rm -f lz4_amalgamated.c

# Generate amalgamated source
echo > lz4_amalgamated.c
cat file1.c >> lz4_amalgamated.c
echo >> lz4_amalgamated.c
cat file2.c >> lz4_amalgamated.c
echo >> lz4_amalgamated.c
# and so on ...
```

the `echo`'s in between take care of newlines between adjacent files.

**Comment by Cyan4973** [comment]
This feature has been partially implemented by @antinucleon, in #600,
enabling amalgamation of `lz4.c` and `lz4hc.c`.

`lz4frame.c` must still be compiled separately.

**Comment by Cyan4973** [comment]
With #610, @antinucleon completed this feature, 
now available on `dev` branch, and will be part of next release.
